connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala

```diff
@@ -218,6 +218,17 @@ private[sql] class AvroSerializer(
         val numFields = st.length
         (getter, ordinal) => structConverter(getter.getStruct(ordinal, numFields))
 
+      case (st: StructType, UNION) =>
+        val unionConvertor = newComplexUnionConverter(st, avroType, catalystPath, avroPath)
```

Comment:
This seems different from https://github.com/apache/spark/blob/master/connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala#L296
If only one non-null field in the union, we will create a converter on the non-null field instead of a complex one.

Reply:
@gengliangwang this branch will only be hit if the resolved avro type is determined to be of the type union.

For optional fields, so a union of a type with `null`.  The [resolution ](https://github.com/apache/spark/pull/36506/files/1d950f55e85f11bfb85d9d57324d7ef7916b4a54#diff-e84c7dfce11df94178029c05a8427668c9dab4bb809ed6e362a39ff043a59f89R369) already deterimined that it was not a union type and will hit the branches another branch in this case.  It is also why `newConverter` is almost always called from a type generated by [`resolveNullableType`](https://github.com/apache/spark/pull/36506/files/1d950f55e85f11bfb85d9d57324d7ef7916b4a54#diff-e84c7dfce11df94178029c05a8427668c9dab4bb809ed6e362a39ff043a59f89R358).

Let me know if this is not clear.
Btw, I missed you question somehow, sorry for the late reply.